Mathlib/Topology/Category/Profinite/Extend.lean

```diff
@@ -0,0 +1,207 @@
+
+section Limit
+
+variable {C : Type*} [Category C] (G : Profinite ⥤ C)
+
+/--
+Given a functor `G` from `Profinite` and `S : Profinite`, we obtain a cone on
+`(StructuredArrow.proj S toProfinite ⋙ toProfinite ⋙ G)` with cone point `G.obj S`.
+
+Whiskering this cone with `Profinite.Extend.functor c` gives `G.mapCone c` as we check in the
+example below.
+-/
+@[simps]
+def cone (S : Profinite) :
+    Cone (StructuredArrow.proj S toProfinite ⋙ toProfinite ⋙ G) where
+  pt := G.obj S
+  π := {
+    app := fun i ↦ G.map i.hom
+    naturality := fun _ _ f ↦ (by
+      have := f.w
+      simp only [const_obj_obj, StructuredArrow.left_eq_id, const_obj_map, Category.id_comp,
+        StructuredArrow.w] at this
+      simp only [const_obj_obj, comp_obj, StructuredArrow.proj_obj, const_obj_map, Category.id_comp,
+        Functor.comp_map, StructuredArrow.proj_map, ← map_comp, StructuredArrow.w]) }
+
+example : G.mapCone c = (cone G c.pt).whisker (functor c) := rfl
+
+/--
+If `c` and `G.mapCone c` are limit cones and the projection maps in `c` are epimorphic,
+then `cone G c.pt` is a limit cone.
+-/
+noncomputable
+def isLimitCone (hc : IsLimit c) [∀ i, Epi (c.π.app i)] (hc' : IsLimit <| G.mapCone c) :
+    IsLimit (cone G c.pt) := (functor_initial c hc).isLimitWhiskerEquiv _ hc'
+
+end Limit
+
+section Colimit
+
+variable {C : Type*} [Category C] (G : Profiniteᵒᵖ ⥤ C)
+
+/--
+Given a functor `G` from `Profiniteᵒᵖ` and `S : Profinite`, we obtain a cocone on
+`(CostructuredArrow.proj toProfinite.op ⟨S⟩ ⋙ toProfinite.op ⋙ G)` with cocone point `G.obj ⟨S⟩`.
+
+Whiskering this cocone with `Profinite.Extend.functorOp c` gives `G.mapCocone c.op` as we check in
+the example below.
+-/
+@[simps]
+def cocone (S : Profinite) :
+    Cocone (CostructuredArrow.proj toProfinite.op ⟨S⟩ ⋙ toProfinite.op ⋙ G) where
+  pt := G.obj ⟨S⟩
+  ι := {
+    app := fun i ↦ G.map i.hom
+    naturality := fun _ _ f ↦ (by
+      have := f.w
+      simp only [op_obj, const_obj_obj, op_map, CostructuredArrow.right_eq_id, const_obj_map,
+        Category.comp_id] at this
+      simp only [comp_obj, CostructuredArrow.proj_obj, op_obj, const_obj_obj, Functor.comp_map,
+        CostructuredArrow.proj_map, op_map, ← map_comp, this, const_obj_map, Category.comp_id]) }
+
+example : G.mapCocone c.op = (cocone G c.pt).whisker (functorOp c) := rfl
+
+/--
+If `c` is a limit cone, `G.mapCocone c.op` is a colimit cone and the projection maps in `c`
+are epimorphic, then `cocone G c.pt` is a colimit cone.
+-/
+noncomputable
+def isColimitCocone (hc : IsLimit c) [∀ i, Epi (c.π.app i)] (hc' : IsColimit <| G.mapCocone c.op) :
+    IsColimit (cocone G c.pt) := (functorOp_final c hc).isColimitWhiskerEquiv _ hc'
+
+end Colimit
+
+end Extend
+
+open Extend
+
+section ProfiniteAsLimit
+
+variable (S : Profinite.{u})
+
+/-- A functor `StructuredArrow S toProfinite ⥤ Profinite` whose limit is isomorphic to `S`. -/
+abbrev diagram' : StructuredArrow S toProfinite ⥤ Profinite :=
```

Comment:
I think it would be better for this diagram to have type `StructuredArrow S toProfinite ⥤ FintypeCat`, and work with the composition with `toProfinite` when you need the diagram to take values in `Profinite`.

Reply:
I was trying to make the naming consistent with `Profinite.diagram` from the `asLimit` file. There we have `Profinite.fintypeDiagram` taking values in `FintypeCat` and `Profinite.diagram` its postcomposition with `toProfinite`. Here, the `Profinite.fintypeDiagram'` would just be `StructuredArrow.proj _ _` so I didn't feel that it deserved a special name. But I can add it as another `abbrev`.